server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/InternalComposite.java

```diff
@@ -392,16 +393,17 @@ static Object formatObject(Object obj, DocValueFormat format) {
         return obj;
     }
 
-    private static class ArrayMap extends AbstractMap<String, Object> {
+    static class ArrayMap extends AbstractMap<String, Object> implements Comparable<Map<String,Object>> {
         final List<String> keys;
-        final List<DocValueFormat> formats;
         final Object[] values;
 
         ArrayMap(List<String> keys, List<DocValueFormat> formats, Object[] values) {
             assert keys.size() == values.length && keys.size() == formats.size();
             this.keys = keys;
-            this.formats = formats;
-            this.values = values;
+            this.values = new Object[values.length];
+            for (int i = 0; i < values.length; i++) {
+                this.values[i] = formatObject(values[i], formats.get(i));
```

Comment:
If you want a comparaison of the raw key you cannot perform the formatting in the ctr. Depending on the requested format the value can be formatted to a string which will not preserve the initial order (dates for instance).